lib/appmap/event.rb

```diff
@@ -71,28 +73,54 @@ def add_schema(param, value, always: false)
             elsif value.respond_to?(:first) && value.first
               param[:properties] = object_properties(value.first)
             end
-          elsif value.respond_to?(:as_json)
-            add_schema param, JSON.parse(value.to_json), always: always
+          else
+            json_value = try_as_json(value) || try_to_json(value) 
+            if value != json_value
+              add_schema param, json_value, always: always
+            end
           end
         end
 
+        def try_as_json(value)
+          value.respond_to?(:as_json) && value.as_json
+        end
+        private_instance_methods :try_as_json
+
+        def try_to_json(value)
+          value.respond_to?(:to_json) && JSON.parse(value.to_json)
+        end
+        private_instance_methods :try_to_json
+
         def object_properties(hash)
           hash = hash.attributes if hash.respond_to?(:attributes)
-          hash = hash.to_h if hash.is_a?(Struct)
+
+          hash = try_to_h(hash)
 
           return unless hash.respond_to?(:each_with_object)
           
-          hash.each_with_object([]) do |entry, memo|
-            key, value = entry
-            memo << {
-              name: key,
-              class: value.class.name,
-            }
-          end
+          hash.map { |k, v| { name: k, class: v.class.name } }
         rescue
           warn $!
         end
 
+        def try_to_h(value)
+          return value unless value.respond_to?(:to_h)
+
+          # Includes such bad actors as Psych::Nodes::Scalar.
+          # Also don't try and hashifiy list-ish things.
+          @unhashifiable_classes ||= Set.new([ Array, Set ])
+
+          return value if @unhashifiable_classes.include?(value.class)
+
+          begin
+            value.to_h
+          rescue
+            # warn "#{value.class}#to_h failed: #{$!.message}"
+            @unhashifiable_classes << value.class
+          end
+        end
+        private_instance_methods :try_to_h
```

Comment:
I don't think `private_instance_methods` does what you think it does ;)

It generally works better to just put `private` before the private section of the class.